### PR TITLE
Use a logger for BZip2 error messages

### LINKED ISSUE #2752: Use a logger for BZip2 warnings
It would be nicer to use a logger - the logger of the communicator that requested this initialization to compress or decompress a message.

_Originally posted by @bernardnormier in https://github.com/zeroc-ice/ice/pull/2750#discussion_r1767116499_

### PR DESCRIPTION
**Pull request by pepone**
Fix #2752

### CODE REVIEW

File: csharp/src/Ice/Internal/BZip2.cs [hunk 1]
```diff
@@ -56,45 +56,52 @@ private enum BzStatus
...
+                string libNames = string.Join(", ", getPlatformNativeLibraryNames()).TrimEnd();
+                bool loaded = false;
+                try
+                {
+                    BZ2_bzLibVersion();
+                    loaded = true;
+                }
+                catch (EntryPointNotFoundException)
+                {
+                    Console.Error.WriteLine($"warning: found {libNames} but entry point BZ2_bzlibVersion is missing.");
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/BZip2.cs:76]: That's where you want to use the logger?

File: csharp/src/Ice/Internal/BZip2.cs [hunk 2]
```diff
@@ -56,45 +56,52 @@ private enum BzStatus
...
-                Console.Error.WriteLine();
+                _isLoaded = loaded;
             }
-            return loaded;
-        });
+            return _isLoaded.Value;
+        }
+    }
+
+    private static readonly object _mutex = new object();
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/BZip2.cs:99]: You should move these fields before the first method.

**Review by externl — APPROVED** [review]

File: csharp/src/Ice/ConnectionI.cs [hunk 3]
```diff
@@ -1816,7 +1816,7 @@ private void initiateShutdown()
             Util.currentProtocol.ice_writeMembers(os);
             Util.currentProtocolEncoding.ice_writeMembers(os);
             os.writeByte(Protocol.closeConnectionMsg);
-            os.writeByte(BZip2.isLoaded ? (byte)1 : (byte)0);
+            os.writeByte(BZip2.isLoaded(_logger) ? (byte)1 : (byte)0);
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/ConnectionI.cs:1819]: In C++ we set this always to 1, in Java to 0

https://github.com/zeroc-ice/ice/blob/5ffe73e10017918fca48cf6aee081c8ec474cfa3/cpp/src/Ice/ConnectionI.cpp#L2290-L2291

https://github.com/zeroc-ice/ice/blob/5ffe73e10017918fca48cf6aee081c8ec474cfa3/java/src/Ice/src/main/java/com/zeroc/Ice/ConnectionI.java#L1603-L1605

[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:1819]: According to https://doc.zeroc.com/ice/3.7/ice-protocol-and-encoding/protocol-compression, we should always use 0 for ValidateConnection and CloseConnection messages.

Given our history of sending "1", we should not enforce this rule when we receive a ValidateConnection or CloseConnection message.

File: csharp/src/Ice/ConnectionI.cs [hunk 4]
```diff
@@ -2232,7 +2232,7 @@ private OutputStream doCompress(OutputStream decompressed, bool compress)
         }
 
         decompressed.pos(9);
-        decompressed.writeByte((byte)((BZip2.isLoaded && compress) ? 1 : 0));
+        decompressed.writeByte((byte)((BZip2.isLoaded(_logger) && compress) ? 1 : 0));
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/ConnectionI.cs:2234]: Added a comment

File: csharp/src/Ice/ConnectionI.cs [hunk 5]
```diff
@@ -2191,48 +2191,47 @@ private int sendMessage(OutgoingMessage message)
...
+                decompressed.writeInt(cstream.size());
 
-                    return cstream;
-                }
+                return cstream;
             }
         }
 
+        // Write the compression status. If BZip2 is loaded and compress is set to true, we write 1, to request a
+        // compressed reply. Otherwise, we write 0 either BZip2 is not loaded or we are sending and uncompressed reply.
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:2232]: an uncompressed

**Review by bernardnormier — APPROVED** [review]
Looks good.

**Pull request merged**